### DISCUSSION TITLE
katz vs eigenvector centrality

### DISCUSSION THREAD

**Discussion by milos-matijasevic** [networkx/networkx#4615] [Q&A]
Hello, i have a question about katz centrality, mostly about parameter beta.

As in the official [katz doc](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.centrality.katz_centrality_numpy.html#networkx.algorithms.centrality.katz_centrality_numpy)
is stated this `When 𝛼=1/𝜆max and 𝛽=0, Katz centrality is the same as eigenvector centrality.`

But then here in katz tests, i see that the same values for katz and eigenvector centrality are expected when (𝛼=1/𝜆max) but `beta=1`, not 0 as quoted in doc.
https://github.com/networkx/networkx/blob/f6007c975384d61c59588d563bf10a8833134211/networkx/algorithms/centrality/tests/test_katz_centrality.py#L341

Also i tried same test but to put beta values to 0. but what i get is this:
```python
import networkx as nx
import numpy as np
G = nx.gnp_random_graph(3, 0.5, seed=1234)
l = float(max(np.linalg.eigvals(nx.adjacency_matrix(G).todense())))
e = nx.eigenvector_centrality(G)
k = nx.katz_centrality_numpy(G, alpha = 1.0 / l, beta = 0)
print(e)
print(k)
```
```
{0: 0.5000000644180599, 1: 0.5000000644180599, 2: 0.707106690085642}
{0: nan, 1: nan, 2: nan}
```

Also if i change beta to any non-negative value i got the same values for k, so i don't see in this example how beta affects katz centrality values.

Can you explain to me what am i missing, what value range for beta is expected and what parameter beta represents?

As i read in the documentation it should be [-1,1] and it represents influence vs authority strength of nodes, is that correct here or there is something else?

**Comment by milos-matijasevic** [comment]
@rossbar @dschult hello, please if you are available I will really appreciate it if you can answer this today or tomorrow

**Comment by rossbar** [comment]
I'm not too familiar with any of the centrality measures, but the docs do seem to be confusing. In the docstring description (and other math environments) it is implied that `beta` is a scalar, but in the implementation beta is treated as a per-node value. The parameter description makes some reference to this in pointing out that `beta` can be a dictionary keyed by node, but that's difficult to reconcile with some of the other references to beta in the docstring.

My guess is that the linked references could clear this up - @milos-matijasevic would you mind taking a look at [2] and seeing if you can help disambiguate things?

**Comment by milos-matijasevic** [comment]
Hey, sorry for the delay, beta parameter that i learned about is explained [here](http://www.leonidzhukov.net/hse/2014/socialnetworks/papers/Bonacich-Centrality.pdf)  in [1] reference in [eigenvector centrality doc](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.centrality.eigenvector_centrality.html#networkx.algorithms.centrality.eigenvector_centrality).
There are more papers i found about explaining beta parameter [paper1](https://www.cmu.edu/joss/content/articles/volume12/Rodan.pdf), [paper2](https://arxiv.org/pdf/1608.05845.pdf).

However, there is another interpretation of beta explained [here](https://www.cs.upc.edu/~CSN/slides/06centrality.pdf) at slide 17, where it says this: `β is the free contribution for
all vertices; hence, no vertex has zero centrality and will contribute
at least β to other vertices centrality.` and i think this is implemented in here.
So if we put beta to be a scalar, every node will add the same weight to all others so normalized Katz centrality will be the same for every beta. (i still don't know why it is none when the beta is 0) and this is why in my test results are the same with every value of beta i tried.
If we pass the dictionary of betas with different values, then the result of the katz_centrality function will change.

So documentation needs to be better i think. And maybe considering implementing beta explained in the first 3 papers.

**Reply by dschult** [reply]
> So if we put beta to be a scalar, every node will add the same weight to all others so normalized Katz centrality will be the same for every beta. (i still don't know why it is none when the beta is 0) and this is why in my test results are the same with every value of beta i tried.

I get the "every node will add the same weight to ..." its neighbors part. But I don't think that means the centrality will "be the same for every beta".  Wouldn't a large beta swamp all information from the eigenvector and thus when normalized all values would be the same?  While if beta is close to zero, the eigenvector part of the process swamps the beta part and you get the eigenvector centrality.  

I also don't understand "why it is none when beta is 0".  Did you mean "nan" instead of "none"?
The value "nan" is a standard floating point representation for "not a number" and usually occurs when dividing zero by zero. Once one value is nan, any arithmetic using that will become nan also. So somehow it is dividing by zero in that example.
